### ISSUE TITLE
Automatically update entry in mixin configuration json when renaming/moving mixin class

### ISSUE THREAD

**Issue by Barteks2x** [minecraft-dev/MinecraftDev#144]
I'm not sure if it's already possible, but would be nice to have a way to add these entries automatically, and warn if they are missing (and possibly show warning if class specified in configuration doesn't exist)

**Comment by DemonWav** [comment]
@Mumfrey thoughts?

**Comment by Mumfrey** [comment]
As a *refactoring* this makes loads of sense.

* Assume you have a mixin `abc.foo.Bar` and rename it to `abc.foo.wibble.Baz`
  1. Find mixin config which declares base package `abc`, or base package `abc.foo`
  2. Find mixin entry for `foo.Bar` (if package match is `abc`) or `Bar` (if package is `abc.foo`) in common, server, or client group
  3. Change entry appropriately: eg. `foo.wibble.Baz` or `wibble.Baz`

So basically if it's been added in the past it's reasonably straightforward to propagate changes when refactoring.

---

*Adding* entries is trickier because:

* There may be multiple mixin sets (configs)
* The same mixin may appear in multiple sets, or not at all (added at runtime via companion)
* Mixin may need to go into common, client, or server group

So to *add* entries you'd need to:
1. Locate available mixin sets
2. Prompt for the set to use if more than one is found
3. Prompt for the group to use: common/client/server

---

*Warning* if a mixin is not present in any sets:

* Not possible, this is totally allowed

---

*Warning* if a set references a mixin which doesn't exist:

* Pretty straightforward, just check if the class exists in the project. Possible extension sideonly functionality here too.

**Comment by Barteks2x** [comment]
I don't quite agree with no warning when mixin is not present in any set, unless you are using companion plugin (and in this case it should be possible to disable that warning), it's almost always a mistake. While it is allowed, it usually makes no sense.

**Comment by Mumfrey** [comment]
I guess if it's only a warning then there's no harm in it. Potentially just suppress the warning if a companion plugin is specified for any set in the same module, and emit the warning if not.

**Comment by Mumfrey** [comment]
The whole point of additive companion plugins though is that you might write two mixins:

`MixinToUseInGeneral`
`MixinToUseIfOptifineDetected`

and have the plugin add the relevant one at runtime based on heuristics. Neither of these mixins would appear in a set and would therefore get a warning attached to them, even though they're perfectly valid.

**Comment by Barteks2x** [comment]
That could be a problem. I've never really worked with companion plugins so I can't be sure if it's possible to handle it. Looking at the documentation, companion plugins are specified in the config json so this analysis can be disabled if companion plugin is found.

**Comment by Mumfrey** [comment]
Companion plugins are quite powerful. They are designed to basically make mixins much more adaptable to different runtime conditions.

**Comment by Barteks2x** [comment]
Yes and there is nothing wrong with using them. But the sanity check could be useful when there is no such plugin. When they are not used, I don't see any reason to have a mixin that isn't used.

**Comment by Mumfrey** [comment]
Indeed, will update the post above to reflect that so that Kyle can use it as a single point of reference.

**Comment by DemonWav** [comment]
Is there any realistic way of determining if a certain Mixin could be defined through a companion plugin using static analysis? A link to an example companion plugin would be really helpful to get a sense of what we're dealing with here.

**Comment by Mumfrey** [comment]
About the best you could do is check whether `getMixins` returns `null` (which it generally will do if the plugin actually has a different purpose). Other than that, `getMixins` is basically allowed to return what it wants under any circumstances that it wants. For example to implement the scenario I mentioned above it might do the following:

```java
@Override
public List<String> getMixins()
{
    Builder<String> list = ImmutableList.<String>builder();
    if (this.detectOptifine())
    {
        list.add("foo.bar.MixinToUseIfOptifineDetected");
    }
    else
    {
        list.add("foo.bar.MixinToUseInGeneral");
        list.add("foo.bar.AnotherMixinWhichAffectsTheSuperClassOfTheOtherOne");
    }

    return list.build();
}
```

or even

```java
@Override
public List<String> getMixins()
{
    String package = String.format("foo.bar.%s.", this.isOptifineDetected() ? "optifine" : "general");
    return ImmutableList.<String>of(
        package + "MixinClassA",
        package + "MixinClassB"
    );
}
```

So in the first example you have a chance of finding the mixin name as a literal string in the method body. In the latter it would be much trickier.

**Comment by DemonWav** [comment]
Yeah, looks like a fool's errand to attempt to figure those out.

**Comment by Mumfrey** [comment]
> Yeah, looks like a fool's errand to attempt to figure those out.

Quite. I mean it could get *more* complicated than that, there's literally nothing stopping the plugin pre-computing the lists, loading the lists for each scenario from a separate config file, you name it. That's the problem with these damned **TURING-COMPLETE COREMODS**. Seriously what is the world coming to where stuff is allowed to be **TURING-COMPLETE**.

**Comment by Mumfrey** [comment]
TURING COMPLETEEE!!!!!

**Comment by DemonWav** [comment]
Mojang/LegacyLauncher/pull/15

**Comment by Minecrell** [comment]
Renaming is implemented in #188
